bundles/org.openhab.binding.salus/src/main/java/org/openhab/binding/salus/internal/rest/AbstractSalusApi.java

```diff
@@ -0,0 +1,114 @@
+import org.openhab.binding.salus.internal.rest.exceptions.AuthSalusApiException;
+import org.openhab.binding.salus.internal.rest.exceptions.SalusApiException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Martin Grześlowski - Initial contribution
+ */
+@NonNullByDefault
+public abstract class AbstractSalusApi<AuthT> implements SalusApi {
+    private static final int MAX_RETRIES = 3;
+    protected static final long TOKEN_EXPIRE_TIME_ADJUSTMENT_SECONDS = 3;
+    protected final Logger logger;
+    protected final String username;
+    protected final byte[] password;
+    protected final String baseUrl;
+    protected final RestClient restClient;
+    protected final GsonMapper mapper;
+    @Nullable
+    protected ZonedDateTime authTokenExpireTime;
+    protected final Clock clock;
+    @Nullable
+    protected AuthT authentication;
+
+    protected AbstractSalusApi(String username, byte[] password, String baseUrl, RestClient restClient,
+            GsonMapper mapper, Clock clock) {
+        this.username = username;
+        this.password = password;
+        this.baseUrl = removeTrailingSlash(baseUrl);
+        this.restClient = restClient;
+        this.mapper = mapper;
+        this.clock = clock;
+        // thanks to this, logger will always inform for which rest client it's doing the job
+        // it's helpful when more than one SalusApi exists
+        logger = LoggerFactory.getLogger(this.getClass().getName() + "[" + username.replace(".", "_") + "]");
+    }
+
+    public AbstractSalusApi(String username, byte[] password, String baseUrl, RestClient restClient,
+            GsonMapper mapper) {
+        this(username, password, baseUrl, restClient, mapper, Clock.systemDefaultZone());
+    }
+
+    protected @Nullable String get(String url, RestClient.Header... headers) throws SalusApiException {
+        refreshAccessToken();
+        return restClient.get(url, headers);
+    }
+
+    protected @Nullable String post(String url, RestClient.Content content, RestClient.Header... headers)
+            throws SalusApiException {
+        refreshAccessToken();
+        return restClient.post(url, content, headers);
+    }
+
+    private static String removeTrailingSlash(String str) {
+        if (str.endsWith("/")) {
+            return str.substring(0, str.length() - 1);
+        }
+        return str;
+    }
+
+    private void forceRefreshAccessToken() throws SalusApiException {
+        logger.debug("Force refresh access token");
+        cleanAuth();
+        refreshAccessToken();
+    }
+
+    protected synchronized final void refreshAccessToken() throws AuthSalusApiException {
+        if (this.authentication == null || isExpiredToken()) {
+            cleanAuth();
+            try {
+                login();
+            } catch (Exception ex) {
+                cleanAuth();
+                throw new AuthSalusApiException("Could not log in, for user " + username, ex);
```

Comment:
In the previous PR we talked about Exception handling. While it has improved, i still some room for improvement. 
In this case, all exceptions are routed to a `AuthSalusApiException` with text about a specific user that cannot be logged in. All is true, but when differentiate between a `AuthSalusApiException` and `SalusApiCommunicationException` it would be easier to handle and inform the user about a possible user/password problem or a transient network failure

Reply:
done